gk: add queues and Solicitor block

cjdoucette opened the pull request:
Add the two types of queues needed for a Gatekeeper server: a
priority request queue that is serviced by a Solicitor block,
and granted queues for each GK instance.

mengxiang0811 commented on sol/main.c:
```diff
@@ -0,0 +1,513 @@
+	}
+
+	/* Insert at head of queue. */
+	if (priority > req_queue->highest_priority) {
```
Does a higher value of priority represent a lower priority? If so, shouldn't one insert the at the head of queue when ``priority < req_queue->highest_priority``?

cjdoucette replied:
A higher value of `priority` represents a higher priority, so if the priority of the packet we're inserting is larger than the highest priority that's currently in the queue (larger than `req_queue->highest_priority`), then we insert it at the head of the queue.

mengxiang0811 commented on sol/main.c:
```diff
@@ -0,0 +1,513 @@
+
+	/* Insert in middle or end of queue. */
+
+	if (priority < req_queue->lowest_priority) {
```
Same situation as above.

cjdoucette replied:
Here we just need to make sure we're keeping the value of `req_queue->lowest_priority` up-to-date.

mengxiang0811 commented on sol/main.c:
```diff
@@ -0,0 +1,513 @@
+		req_enqueue(sol_conf, req_nodes[i]);
+}
+
+/* Scaling for cycles_per_byte calculation
```
The first line of the comment should be ``/*``. It needs a period at the end of the comment.

mengxiang0811 commented on sol/main.c:
```diff
@@ -0,0 +1,513 @@
+/* Scaling for cycles_per_byte calculation
+ * Chosen so that minimum rate is 480 bit/sec
+ */
+#define RTE_SCHED_TIME_SHIFT		      8
```
Need a parentheses for the number ``8``.

mengxiang0811 commented on sol/main.c:
```diff
@@ -0,0 +1,513 @@
+
+		/* Check credits. */
+		if (!credits_check(req_queue, pkt)) {
+			printf("out of credits\n");
```
Use ``RTE_LOG()`` instead of printf's.

mengxiang0811 commented on sol/main.c:
```diff
@@ -0,0 +1,513 @@
+
+		/* Remove extra space in mbuf. */
+		if (rte_pktmbuf_adj(pkt, sizeof(*head)) == NULL) {
+			printf("bug in request queue: should be able to remove the extra linked list data in a packet, but the removal failed\n");
```
Use ``RTE_LOG()`` instead of printf's.

mengxiang0811 commented on sol/main.c:
```diff
@@ -0,0 +1,513 @@
+		req_queue->highest_priority = head->priority;
+
+	do {
+		sent = rte_eth_tx_burst(tx_port, tx_queue,
```
The ``sent`` can be defined here, since it's only used within this block.

cjdoucette replied:
The value of `sent` needs to persist across iterations of the loop (it's used in the expression `pkts_out + sent` here), so we need to declare it up top and update it during every iteration of the loop.

mengxiang0811 replied:
I see. So, shouldn't the expression be ``pkts_out + tot_sent``? That is an accumulated value of number of packets that have been sent out? However, ``sent`` here is the number of packets that were sent out successfully for one iteration of the loop. ``nb_pkts_out`` should be the number of remaining unsent packets.

mengxiang0811 commented on sol/main.c:
```diff
@@ -0,0 +1,513 @@
+		req_queue->len--;
+
+		/* Remove extra space in mbuf. */
+		if (rte_pktmbuf_adj(pkt, sizeof(*head)) == NULL) {
```
Why does one need to remove ``sizeof(*head)`` space in ``mbuf``? It seems that ``mbuf`` is the packet received from NIC without any change.

cjdoucette replied:
Right, that was left over from the last implementation. Thanks!

cjdoucette commented on sol/main.c:
```diff
@@ -0,0 +1,513 @@
+		goto out;
+	}
+
+	ret = net_launch_at_stage1(net_conf, 0, 0, 1, 0, sol_state1, sol_conf);
```
`sol_stage1()` won't run until sometime after `run_sol()` is finished, so I think it should be OK to set `sol_conf->net` later in this function.

cjdoucette commented:
Thanks for the comments, Qiaobin. I added your suggestions, but I'm going to go one step further and replace the linked list I'm using here with the linked list we already have in include/list.h. I'll let you know when that's ready.

cjdoucette commented:
I added more comments and am also now using the `struct list_head` implementation for the priority queue. Let me know if anything is unclear!

mengxiang0811 commented on sol/main.c:
```diff
@@ -0,0 +1,449 @@
+	struct sol_config *sol_conf = arg;
+	int ret;
+
+	ret = get_queue_id(&sol_conf->net->back, QUEUE_TYPE_TX,
```
It's better to define ``ret`` this line.

mengxiang0811 commented on sol/main.c:
```diff
@@ -0,0 +1,449 @@
+	/* Insert in middle or end of queue. */
+	if (priority < req_queue->lowest_priority)
+		req_queue->lowest_priority = priority;
+	for (next = priority + 1; next < req_queue->num_priorities; next++) {
```
Isn't ``for (next = priority + 1; next <= req_queue->highest_priority; next++) {`` enough here?

mengxiang0811 commented on sol/main.c:
```diff
@@ -0,0 +1,449 @@
+
+	if (req_queue->len == sol_conf->pri_req_max_len) {
+		/* New packet is lowest priority, so drop it. */
+		if (req_queue->lowest_priority == priority) {
```
Shouldn't it be ``req_queue->lowest_priority >= priority``? Since if the priority is smaller than the lowest one, it will be dropped as well.

cjdoucette replied:
Yes, that's true. Thank you!

cjdoucette commented:
OK, this pull request is ready for another round of reviews. I added another patch with RED queues to the pull request as well.

cjdoucette commented:
I just added the code to set the ECN bits in the IPv4 and IPv6 headers when congestion is found, so these two commits are ready for review.

cjdoucette commented:
I rebased the last commit to include code that recomputes the inner header's checksum when the ECN bit is changed.

I also considered the option of just zeroing out the inner header's checksum field. That way, when Grantor decapsulates and sends the packet, the IPv4 checksum will be written by hardware on transmission from Grantor. If we can make the assumption that Grantor nodes will have this hardware feature, then I can remove this recomputation.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,459 @@
+	ret = init_mailbox("gk_reqs", 2 * sol_conf->pri_req_max_len,
+		sizeof(struct priority_req), sol_conf->lcore_id,
+		&sol_conf->mb);
+	if (ret < 0)
```
Style: I think `init_mailbox()` should be in `run_sol()`. I don't get why it's postponed to stage 1.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,459 @@
+
+	while (likely(!exiting)) {
+		enqueue_reqs(sol_conf);
+		dequeue_reqs(sol_conf, tx_port_back,
```
This is not a bug, neither an issue. I forgot that `dequeue_reqs()` is throttled to 5% of the bandwidth.

Ignore this comment.

AltraMayor commented on include/gatekeeper_sol.h:
```diff
@@ -0,0 +1,122 @@
+	/* TX queue on the back interface. */
+	uint16_t           tx_queue_back;
+	struct net_config  *net;
+	struct gk_config   *gk;
```
The field `gk` is never used. Moreover, it's not the Solicitor block that depends on the GK blocks, but the other way around. The correct dependency should be reflected in the configuration files because if a GK block is running and the Solicitor is not, the code will crash.

cjdoucette commented on lua/sol.lua:
```diff
@@ -0,0 +1,21 @@
+	sol_conf.lcore_id = lcore
+	sol_conf.pri_req_max_len = 1024
+	sol_conf.enq_burst_size = 1024
+	sol_conf.deq_burst_size = 1024
```
Don't quite understand this comment -- are you saying that `enq_burst_size` and `deq_burst_size` should be removed and we should just use `GATEKEEPER_MAX_PKT_BURST`? Or that the code should make sure they are equal? Or something else?

AltraMayor replied:
I just mean that the values we are setting are too big. Set `sol_conf.enq_burst_size` and `sol_conf.deq_burst_size` with the same value of `GATEKEEPER_MAX_PKT_BURST` for now. We'll think of optimal values later.

cjdoucette commented on sol/main.c:
```diff
@@ -0,0 +1,459 @@
+	for (next = priority + 1; next < req_queue->highest_priority; next++) {
+		if (req_queue->priorities[next] != NULL) {
+			list_add(&pr->list, &req_queue->priorities[next]->list);
+			return;
```
It can't be the case that `priority == req_queue->highest_priority` because this function only inserts a request for a priority that does not currently exist in the queue. But I will add an `RTE_VERIFY()` to make sure this expectation is met.

cjdoucette commented on sol/main.c:
```diff
@@ -0,0 +1,459 @@
+	if (pkt_len > req_queue->tb_credits)
+		return false;
+
+	req_queue->tb_credits -= pkt_len;
```
In that case, the `if` statement above will catch that there are not enough credits and we will return.

AltraMayor replied:
My fault. Thanks.

cjdoucette commented on sol/main.c:
```diff
@@ -0,0 +1,459 @@
+ * Return an error if trying to allocate the second instance.
+ */
+struct sol_config *
+alloc_sol_conf(void)
```
Since the Solicitor block is only needed for Gatekeeper servers, I chose to allocate it dynamically (same as the GK-GT Unit block).

cjdoucette commented on sol/main.c:
```diff
@@ -0,0 +1,459 @@
+	req_queue->time_cpu_bytes = 0;
+	req_queue->time = 0;
+
+	INIT_LIST_HEAD(&req_queue->head);
```
Technically no, since we are allocating the entire configuration structure (which includes `req_queue`) using `rte_calloc()`, which zeroes the memory. Still, I've explicitly set some fields to zero in this initialization function, so I will zero the `priorities` array to make it consistent.

AltraMayor replied:
Since we're going to have set all fields, swap `rte_calloc()` to a regular `rte_malloc()`.

cjdoucette commented:
All suggestions have been addressed either in the comments above or the latest rebase.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+	}
+
+	next_lowest_pr = list_last_entry(&lowest_pr->list,
+		struct priority_req, list);
```
It works, but we really mean `list_prev_entry(&lowest_pr, list)` here. Please add `list_next_entry()` and `list_prev_entry()` to `include/list.h`.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+drop:
+	list_del(&lowest_pr->list);
+	mb_free_entry(&sol_conf->mb, lowest_pr);
+	rte_pktmbuf_free(lowest_pr_pkt);
```
Variable `lowest_pr_pkt` is only needed on this line. If you move `rte_pktmbuf_free()` before `mb_free_entry()`, we don't need variable `lowest_pr_pkt`.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+	struct req_queue *req_queue)
+{
+	struct priority_req *lowest_pr =
+		req_queue->priorities[req_queue->lowest_priority];
```
Stype: the expression `list_last_entry(&req_queue->head, struct priority_req, list);` to obtain `lowest_pr` seems simpler and more inline with the rest of the code in this function.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+		if (req_queue->lowest_priority >= priority) {
+			struct rte_mbuf *lowest_pri_pkt = req->pkt;
+			mb_free_entry(&sol_conf->mb, req);
+			rte_pktmbuf_free(lowest_pri_pkt);
```
If you move `rte_pktmbuf_free()` before `mb_free_entry()`, we don't need variable `lowest_pri_pkt`.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+		req_queue->time_bytes += pkt->pkt_len +
+			RTE_SCHED_FRAME_OVERHEAD_DEFAULT;
+
+		if (nb_pkts_out == sol_conf->deq_burst_size)
```
Style: `==` is correct, but `>=` is robust.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+			pkts_out + total_sent, nb_pkts_out);
+		total_sent += sent;
+		nb_pkts_out -= sent;
+	} while (nb_pkts_out > 0);
```
It's not a bug, but this should be a `while` loop instead of a `do..while` loop because `nb_pkts_out` might be zero.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+	while (likely(!exiting)) {
+		enqueue_reqs(sol_conf);
+		dequeue_reqs(sol_conf, tx_port_back,
+			sol_conf->tx_queue_back);
```
Drop the third parameter of `dequeue_reqs()`; `dequeue_reqs()` can read `sol_conf->tx_queue_back` thanks to its first parameter. There's no gain in having the third parameter.

The second parameter is okay.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+
+	ret = net_launch_at_stage1(net_conf, 0, 0, 0, 1, sol_stage1, sol_conf);
+	if (ret < 0)
+		goto out;
```
We need to release the mailbox here.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+		sol_conf = rte_malloc("sol_config",
+			sizeof(struct sol_config), 0);
+		return sol_conf;
+	} else {
```
Style: we don't need the `else` clause.

AltraMayor commented on gk/main.c:
```diff
@@ -415,6 +427,14 @@ gk_process_granted(struct flow_entry *fe, struct ipacket *packet)
 	return 0;
```
This dropped comment should become part of the explanation of returning zero in the previous item of this review.

AltraMayor commented on gk/main.c:
```diff
@@ -917,20 +937,25 @@ process_pkts_front(uint8_t port_front, uint8_t port_back,
 
 		if (ret < 0)
 			rte_pktmbuf_free(pkt);
```
It would be worth replacing all calls to `rte_pktmbuf_free()` with `drop_packet()` in the GK blocks. This should help us when we start to collect statistics.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+{
+	static rte_atomic16_t num_sol_conf_alloc = RTE_ATOMIC16_INIT(0);
+	if (rte_atomic16_test_and_set(&num_sol_conf_alloc) == 1) {
+		sol_conf = rte_malloc("sol_config",
```
1. We need to bring `rte_calloc()` back here, otherwise `cleanup_sol()` may fail when it's called while Gatekeeper is still initializing. Sorry or having asked to change it to `rte_malloc()`; a comment explaining it would guarantee that nobody is going to consider the same mistake again.

2. Before returning, call `INIT_LIST_HEAD(&req_queue->head);` to guarantee that `cleanup_sol()` won't fail for the same reason above.

AltraMayor commented on include/gatekeeper_sol.h:
```diff
@@ -0,0 +1,118 @@
+	/* Number of credits accrued for every token bucket period. */
+	uint32_t              tb_credits_per_period;
+	/* Capacity of the token bucket (the max number of credits). */
+	uint32_t              tb_size;
```
Shouldn't field `tb_size` be called `tb_size_credits` to incorporate the unit?

AltraMayor commented:
Hi @cjdoucette,

Please create another pull request for the second patch, so we can merge the first patch sooner. Once we merge the first patch, you can create a pull request for the VLAN ID support before we merge the pull request for the second patch.

cjdoucette commented on lua/gatekeeper_config.lua:
```diff
@@ -22,10 +22,13 @@ function gatekeeper_init()
 	if gatekeeper_server == true then
 		local gkf = require("gk")
-		local gk_conf, ggu_lcore = gkf(net_conf, numa_table)
+		local gk_conf, ggu_lcore, sol_lcore = gkf(net_conf, numa_table)
```
OK, although this forces the declaration of `n_lcores` out of `lua/gk.lua` and into `lua/gatekeeper_config.lua`.

cjdoucette commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+static inline int
+credits_check(struct req_queue *req_queue, struct rte_mbuf *pkt)
+{
+	uint32_t pkt_len = pkt->pkt_len + RTE_SCHED_FRAME_OVERHEAD_DEFAULT;
```
That constant defines the Ethernet framing overhead as bytes that count toward the number of credits used. Its definition is:

    /*
     * Ethernet framing overhead. Overhead fields per Ethernet frame:
     * 1. Preamble:                             7 bytes;
     * 2. Start of Frame Delimiter (SFD):       1 byte;
     * 3. Frame Check Sequence (FCS):           4 bytes;
     * 4. Inter Frame Gap (IFG):               12 bytes.
     *
     * The FCS is considered overhead only if not included in the packet
     * length (field pkt_len of struct rte_mbuf).
     */
    #ifndef RTE_SCHED_FRAME_OVERHEAD_DEFAULT
    #define RTE_SCHED_FRAME_OVERHEAD_DEFAULT      24
    #endif

I'll add a comment for this.

cjdoucette commented on sol/main.c:
```diff
@@ -0,0 +1,507 @@
+
+	/* Calculate link speed in bytes per second. */
+	rte_eth_link_get(sol_conf->net->back.id, &link);
+	if (link.link_speed == ETH_SPEED_NUM_NONE) {
```
It actually looks like it does NOT behave that way -- it will just return the link speed of the first slave. Thanks for pointing this out, as I will need to sum the link speeds of all the slaves in that case.

cjdoucette commented:
I just rebased the pull request with all of the changes and with only the first commit.

I was able to simplify the token bucket algorithm according to what we discussed. I think some of the variables that were in place before were used to keep track of time in terms of cycles and bytes that could then be used by multiple entities for separate token buckets. However, since our usage only has one token bucket (for the request queue), these variables were holding the same values, so I got rid of them.

I also added a feature that saves unused cycles for future iterations of the credits computation. Otherwise, these cycles disappear because of rounding and can't be used. This wasn't present before, but should give a more accurate count of the number of available bytes.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,478 @@
+	uint64_t avail_bytes;
+
+	/* Not enough cycles have passed to update the number of credits. */
+	/* XXX We should run tests to see if this is (un)likely. */
```
We can drop this `XXX` comment because not only does the probability depend on our code, but also on how fast the processor and NICs are.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,478 @@
+	}
+	link_speed_bytes = ((uint64_t)link.link_speed * 1000 * 1000 / 8);
+	req_queue->cycles_per_byte =
+		rte_get_tsc_hz() / (sol_conf->req_bw_rate * link_speed_bytes);
```
1. We'll still need `spare_cycles`, but there's a small trick to eliminate it as a field: decrementing the spare cycles from `time_cpu_cycles` while updating `time_cpu_cycles`.

2. Given that we need the spare cycles, `rte_reciprocal_*()` is not that useful because it implies two multiplications: one to obtain bytes from cycles, and another to obtain the spare cycles. We can drop `rte_reciprocal_*()` and use `ldiv()` to do both at the cost of a division.

3. Since we're going to save `cycles_per_byte` (the double) as a fraction, suggestion for field names: `cycles_per_byte_a` and `cycles_per_byte_b`.

cjdoucette commented:
This rebase contains all the suggestions from above.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,551 @@
+gk_solicitor_enqueue(struct sol_config *sol_conf, struct rte_mbuf *pkt,
+	uint8_t priority)
+{
+	struct priority_req *req_node = mb_alloc_entry(&sol_conf->mb);
```
We have to test that `priority` is in [0, 63] because the code will break if this is false. If `priority` is not valid, the function should log a message and return an error.

AltraMayor commented on include/gatekeeper_sol.h:
```diff
@@ -0,0 +1,131 @@
+};
+
+/* Number of priorities available in the IP DSCP field. */
+#define GK_NUM_REQ_PRIORITIES (64)
```
Almost everywhere `GK_NUM_REQ_PRIORITIES` is being used, it's meant to mean 63 instead of 64. I suggest renaming this constant to `GK_MAX_REQ_PRIORITY` and adding plus one where needed. Don't forget to adjust the comment.

cjdoucette commented on sol/main.c:
```diff
@@ -0,0 +1,551 @@
+		}
+
+		if (req_queue->len == 1 || (entry->priority != next->priority))
+			req_queue->priorities[entry->priority] = NULL;
```
If `entry` is the lowest priority packet, then `next` is the `list_entry()` of the head of the list, which is not a valid pointer to a `struct priority_req`.

Putting the test `req_queue->len == 1` first is enough to protect against `next` being used when it's not a valid pointer, but I could see how that could be perceived as sloppy. I can't find any examples of using `list_for_each_entry_safe()` where the temporary storage parameter (`next` in this case) is dereferenced, perhaps for this reason.

Should I instead keep a variable like `prev_priority` and update it in the loop to do an equivalent test without dereferencing `next`?

AltraMayor replied:
Let's keep the way it's now. Thank you.

cjdoucette commented:
The only suggestion I didn't address is the one I commented on above.

AltraMayor commented on sol/main.c:
```diff
@@ -0,0 +1,568 @@
+	if (unlikely(req_queue->len == 1)) {
+		req_queue->priorities[lowest_pr->priority] = NULL;
+		req_queue->highest_priority = 0;
+		req_queue->lowest_priority = GK_MAX_REQ_PRIORITY + 1;
```
I understand that this and similar cases don't need the plus one, otherwise we shouldn't have zero in `highest_priority`.

cjdoucette replied:
I see -- agreed, I will drop the `+ 1`.
